### ISSUE TITLE
Clarification - Converters vs on_setattr / Convert automatically when setting 

### ISSUE THREAD

**Issue by danieldjewell** [python-attrs/attrs#835]
attrs version: 21.2.0

Using the documentation [on Converters](https://www.attrs.org/en/stable/init.html#converters) it was not immediately clear that converters only work on instantiation of a new attrs-enhanced class. (Or I totally missed something.) 

1. Is this the intended function? That the type conversion only happens on class instantiation and not when setting an attribute? (e.g. ```o.x = "2" ``` in the example in the documentation)
    * If yes: Perhaps the documentation could be clarified to explain this and also provide a clear example of using ```on_setattr``` to also run the converter
    * If no: Perhaps there's a bug because setting an attribute when a converter is specified does not work

And further: from a logical design standpoint, it seems that if you're specifying a converter to be run at class instantiation [which means *all* values on *new* instances are run through that converter....] what circumstances would you *not* want that to run when setting attributes? Part of the purpose of converters is to ensure consistent typing and normalization (from the documentation: _"Finally, sometimes you may want to normalize the values coming in. For that attrs comes with converters."_). Given that, it seems strange that you could break that normalization by simply setting an attribute.

TL;DR: Converters only run on class instantiation. Shouldn't they be run when setting an attribute as well?

**Comment by wsanchez** [comment]
@danieldjewell You want to learn about [`on_setattr`](https://www.attrs.org/en/stable/api.html?highlight=on_setattr#setters).

When I type `on_setattr` into the [attrs.org](https://www.attrs.org/) search bar, the only useful thing I see for this question is the above link to utilities for `on_setattr`, and there doesn't seem to be mention of it in the [Initialization](https://www.attrs.org/en/stable/init.html) page (eg. under Validation).

So I'll mark this as a documentation issue.

**Comment by wsanchez** [comment]
Marking as documentation because `on_setattr` seems to lack narrative docs, and running converters and validators on set as well as init time is probably a common enough use case to warrant some text.

**Comment by hynek** [comment]
I'm not sure where to put narrative information…should we open a new chapter for it?

**Comment by bitranox** [comment]
@hynek 
I am also complaining ! ;-)
I would definitely need the converters also on_setattr !

```
@attr.define(slots=True)
class MyDataClass(object):
    my_value=attr.ib(default=0, converter=int)

>>> my_data=MyDataClass()
>>> my_data.my_value='1'
>>> assert 1 == my_data.my_value
```
better yesterday then tomorrow ;-),  i just needed to write ugly and slow own converters - 
it would have saved me a lot of effort it that would work.

I just came here to put I question wha its not working - and landed at this thread.

yours sincerely, und Grüße nach Berlin

bitranox, Vienna

**Comment by hynek** [comment]
please see #886
